One Lolly CTL for World of Padman

kai-li-wop opened the pull request:
This is currently a draft and work in progress to try to implement **One Lolly Capture The Lolly (1LCTL)** as a new game type to World of Padman. Most of the code is from Quake3 Team Arena, the assets are original WoP assets but adapted to work with 1LCTL or newly created. Some of it is also just a placeholder. Since Team Arena works a bit different and WoP does not use or contain the Team Arena code, some things need to be done differently or recreated. So help and support is more than welcome. This could be released as a modification for the upcoming World of Padman release or as part of it if accepted by the team.

1LCTL uses the default CTL HUD at the lower right but in a modified way. The upper lolly slot show a blue or red lolly icon (@) depending on the neutral lolly taken status to indicate which team has currently taken the neutral lolly. The lower lolly slot shows different white lolly icons depending on the neutral lolly status; at base (@), taken (x) and dropped (?). I also experimented with only one lolly slot for 1LCTL, but I disliked the outcome.

ToDo
- [x] Transfer and adjust 1FCTF code from Q3TA
- [x] Adjust the CTL HUD to work with 1LCTL
- [x] Provide green lolly icons and shader for the neutral lolly
- [x] Adjust Team Overlay and Scoreboard for 1LCTL 
- [x] Provide a lolly model with white textures
- [ ] Provide necessary sound samples --> #19
- [x] Provide a test map (test with Harm's Westernset)
- [x]  Provide a map source file in XTRAS folder
- [x] Adjust bot AI for 1LCTL --> #261
- [x] Adjust WoP's CTL maps for 1LCTL (except ENTE's maps)

kai-li-wop committed:
1LCTL: initial and work in progress

kai-li-wop committed:
UI: Add 1LCTL to server browser list

mgerhardy commented on code/game/g_team.c:
```diff
@@ -643,32 +664,35 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
 
-	if (cl->sess.sessionTeam == TEAM_RED) {
-		enemy_flag = PW_BLUEFLAG;
+	if(g_gametype.integer == GT_1FCTF) {
```
missing whitespace

mgerhardy commented on code/game/g_team.c:
```diff
@@ -643,32 +664,35 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
 		return 0; // We don't have the flag
-	PrintMsg(NULL, "%s" S_COLOR_WHITE " captured the %s' lolly!\n", cl->pers.netname, TeamName(OtherTeam(team)));
-
+	if( g_gametype.integer == GT_1FCTF ) {
```
formatting is off

mgerhardy commented on code/game/g_team.c:
```diff
@@ -728,14 +752,28 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
-		cl->ps.powerups[PW_BLUEFLAG] = INT_MAX; // flags never expire
+		cl->ps.powerups[PW_NEUTRALFLAG] = INT_MAX; // flags never expire
+
+		if( team == TEAM_RED ) {
```
formatting

mgerhardy commented on code/game/g_items.c:
```diff
@@ -728,11 +728,29 @@ void G_CheckTeamItems(void) {
+	}
+
+	if( g_gametype.integer == GT_1FCTF ) {
+		gitem_t	*item;
```
this must be a `const` pointer

mgerhardy commented on code/game/g_items.c:
```diff
@@ -728,11 +728,29 @@ void G_CheckTeamItems(void) {
+		gitem_t	*item;
+
+		// check for all three flags
+		item = BG_FindItem( "Red Lolly" );
```
formatting is off here

kai-li-wop committed:
JANITORIAL: const and whitespaces

kai-li-wop committed:
WOP: 1LCTL help and info screens

kai-li-wop committed:
GAME: Allow BAMBAM and BOOMiES also in 1LCTL

kai-li-wop committed:
CGAME: Reposition lagometer and disconnect also for 1LCTL

kai-li-wop committed:
CGAME: Register 1LCTL hud bg separate

mgerhardy commented on code/botlib/be_ai_goal.c:
```diff
@@ -89,7 +89,8 @@ typedef struct campspot_s {
 	BOTLIB_GT_FFA,
-	BOTLIB_GT_TEAM
+	BOTLIB_GT_TEAM,
+	BOTLIB_GT_1FCTF,
```
please remove the trailing `,` after `BOTLIB_GT_1FCTF`

mgerhardy commented on code/cgame/cg_draw.c:
```diff
@@ -1323,7 +1323,7 @@ static void CG_DrawDisconnect(void) {
 	} else if (cgs.gametype == GT_BALLOON) {
 		x = (640 - 48 - 64);
-	} else if (cgs.gametype == GT_CTF) {
+	} else if ((cgs.gametype == GT_CTF) || (cgs.gametype == GT_1FCTF)) {
```
`if (cgs.gametype == GT_CTF || cgs.gametype == GT_1FCTF)` this is enough - no extra `()` needed

mgerhardy commented on code/cgame/cg_main.c:
```diff
@@ -1047,7 +1062,16 @@ static void CG_RegisterGraphics(void) {
 		}
 
-		if ((cgs.gametype == GT_CTF) || (cgs.gametype == GT_BALLOON) || cg_buildScript.integer) {
+		if ( cgs.gametype == GT_1FCTF || cg_buildScript.integer ) {
```
formatting is wrong here.

mgerhardy commented on code/cgame/cg_predict.c:
```diff
@@ -291,6 +291,12 @@ static void CG_TouchItem(centity_t *cent) {
 
 	// Special case for flags.
 	// We don't predict touching our own flag
+	if(cgs.gametype == GT_1FCTF) {
```
formatting is wrong here

mgerhardy commented on code/game/bg_misc.c:
```diff
@@ -776,6 +790,22 @@ qboolean BG_CanItemBeGrabbed(int gametype, const entityState_t *ent, const playe
 	case IT_TEAM: // team items, such as flags
+		if (gametype == GT_1FCTF) {
+			// neutral flag can always be picked up
+			if(item->giTag == PW_NEUTRALFLAG) {
```
formatting is wrong here `if(` should be `if (`

mgerhardy commented on code/game/bg_misc.c:
```diff
@@ -776,6 +790,22 @@ qboolean BG_CanItemBeGrabbed(int gametype, const entityState_t *ent, const playe
+				return qtrue;
+			}
+			if (ps->persistant[PERS_TEAM] == TEAM_RED) {
+				if (item->giTag == PW_BLUEFLAG  && ps->powerups[PW_NEUTRALFLAG]) {
```
formatting is wrong here - `  && ` should be ` && `

mgerhardy commented on code/game/g_items.c:
```diff
@@ -728,11 +728,29 @@ void G_CheckTeamItems(void) {
+		}
+	}
+
+	if(g_gametype.integer == GT_1FCTF) {
```
missing whitespace between if and (

mgerhardy commented on code/game/g_team.c:
```diff
@@ -520,6 +533,10 @@ static void Team_ResetFlags(void) {
 		Team_ResetFlag(TEAM_RED);
 		Team_ResetFlag(TEAM_BLUE);
 	}
+
```
else if should be in the same line as the closing brace }

mgerhardy commented on code/game/g_team.c:
```diff
@@ -728,14 +752,28 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
+
+		if (team == TEAM_RED) {
+			Team_SetFlagStatus(TEAM_FREE, FLAG_TAKEN_RED);
+		}
```
else should be in the same line as the closing }

mgerhardy reviewed (commented):
nice job so far - mostly about cosmetic formatting stuff. this is only "important" because the next clang-format run would produce a lot of diffs otherwise.

kai-li-wop commented:
Most of the stuff is simply copy paste from Team Arena and the formatting issues of TA code are not easy to spot. Is there a VS Code plugin for this? Any suggestions?

kai-li-wop committed:
JANITORIAL: Fix minor code formatting issues

kai-li-wop committed:
GAME: Spawn blue and red lollies also in 1LCTL ...

and remove the neutral lolly when not in 1LCTL. Regarding the deleted comment: yes, It is better to hard code those things if mappers forget to set the correct flags.

kai-li-wop committed:
GAME: Add neutral lolly to q3ToWopItems list for q3a map support

kai-li-wop committed:
CGAME: Draw modified HUD lower right for 1LCTL

1LCTL uses CTL HUD bg graphics. The blue/red lolly icon is drawn in upper slot depending on neutral flag taken status to indicate which team has currently taken the neutral lolly. The neutral lolly icon is drawn in lower icon slot depending on neutral flag status: at base, taken, dropped.

kai-li-wop committed:
1LCTL: Providing a small test map

kai-li-wop committed:
GAME: Fix wrong classname

kai-li-wop committed:
ASSETS: 1LCTL demo map and source file

kai-li-wop committed:
CGAME: Add neutral lolly icon to team overlay

kai-li-wop committed:
CGAME: Add neutral lolly to the scoreboard

kai-li-wop committed:
CGAME: Add neutral lolly model and glow to player model

kai-li-wop committed:
GAME: Enable the player also to drop the neutral lolly

kai-li-wop committed:
GAME: Update neutral lolly status if dropped

kai-li-wop committed:
GAME: Remove also neutral lolly when arget_remove_powerups

kai-li-wop committed:
CGAME: Fix playing taken lolly sounds when in 1LCTL

kai-li-wop committed:
CGAME: Also check for neutral lolly pickup even if announcer sound is not played in WoP

kai-li-wop committed:
GAME: Add neutral lolly also to old use item code

... WoP seems not to use this code right now and drops lollies via console cmds.

kai-li-wop committed:
GAME: Check almost capture also for neutral lolly

Probably currently not used by WoP

kai-li-wop committed:
GAME: Suicide also returns the neutral lolly

kai-li-wop committed:
GAME: Prevent BamBam/Boomies being planted too close to neutral lolly

kai-li-wop committed:
CGAME: Play capture/stolen lolly sounds also in 1LCTL

kai-li-wop committed:
CGAME: Omit unnecessary parentheses

kai-li-wop committed:
CGAME: Remove new team arena flag animations code

WoP does not use it as far as I can see and always adds the lollies as trail items. Also the player models do not provide a tag for the flag and the handles for flag models etc are not not in use.

kai-li-wop committed:
CGAME: Flag to Lolly and sound path adjustments

kai-li-wop committed:
CGAME: Fix accidentally deleted almost capture sound

kai-li-wop committed:
CGAME: Fix sound paths and ToDo

kai-li-wop committed:
CGAME: Fix AlmostCapture sound path again due to changes from main

kai-li-wop committed:
GAME: Fix vote menu not working for 1LCTL ...

and all game types following in the list.

kai-li-wop committed:
GAME/CGAME/WOP: Better use white for neutral lolly

Better use the same color scheme like in Big Balloon

kai-li-wop committed:
WOP/GAME: Replace white lolly with neutral one ...

... and change from green to neutral color. The neutral lolly is still a placeholder derived from the blue version until a new model is made by doomdragon. I propose the flavor "vanilla" with a light yellow-brown shade for the neutral lolly design.

kai-li-wop committed:
GAME: Also call Team_FlagDefendOrProtectBonus when 1FCTF is enabled

kai-li-wop committed:
WOP: Enable GlowStar's ColorStage for 1lCTL in map menu

kai-li-wop committed:
WOP: Delete the 1LCTL demo map, test with Westernset or ColorStage instead

kai-li-wop committed:
WOP: Enable CybJourney and FridgeWars maps for 1lCTL in map menu

kai-li-wop committed:
GAME: Add onelolly to G_ValueIncludesGametype to support Q3 maps

kai-li-wop committed:
CGAME: Show medals in scoreboard also in 1LCTL
